eng/Versions.props

```diff
@@ -83,7 +83,7 @@
     <SystemNetWebSocketsWebSocketProtocolPackageVersion>4.6.0</SystemNetWebSocketsWebSocketProtocolPackageVersion>
     <SystemReflectionMetadataPackageVersion>1.7.0</SystemReflectionMetadataPackageVersion>
     <SystemRuntimeCompilerServicesUnsafePackageVersion>4.6.0</SystemRuntimeCompilerServicesUnsafePackageVersion>
-    <SystemSecurityCryptographyCngPackageVersion>4.6.0</SystemSecurityCryptographyCngPackageVersion>
+    <SystemSecurityCryptographyCngPackageVersion>4.6.1</SystemSecurityCryptographyCngPackageVersion>
```

Comment:
@mmitche Why did this change? This is causing the test failures since we didn't expect this package to rev.

Reply:
We built that package as part of 3.0.1, the dependency should've gotten updated back then: https://github.com/dotnet/corefx/pull/41795. Don't know why it didn't rev until now. How do we react?

Reply:
Different versions of System.Security.Cryptography.Cng.dll are landing in _both_ Microsoft.NetCore.App (product v3.0.3.0, file v4.700.20.6701) and Microsoft.AspNetCore.App (product v3.0.1.0, file v4.700.19.51609). If ASP.NET Core was receiving the same version, it wouldn't include the assembly in Microsoft.AspNetCore.App.

Put another way, package version **4.6.1 is incorrect**; we need the System.Security.Cryptography.Cng containing the 3.0.3.0 assembly.

Reply:
4.6.1 is correct here. That's the latest, from 3.0.1, the last time that this was serviced. I checked the input corefx feeds for this package version. It would be coming from either nuget.org or https://pkgs.dev.azure.com/dnceng/public/_packaging/darc-pub-dotnet-corefx-c1778515/nuget/v3/index.json. Both have the same package.

The question is why core-setup appears to have a version that was built in 2020. I'm rerunning the core-setup pipeline to see whether it maybe somehow had an old version as input, but seems unlikely because I don't think a version was ever produced in 2020.

Reply:
There's no 4.6.* version produced in 2020 that darc knows about (as the file version implies)

Reply:
Alright, with the new build, the output was the same from core-setup.  Still the newer Cng.  It's coming from https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet3-transport&package=runtime.win-x64.Microsoft.Private.CoreFx.NETCoreApp&protocolType=NuGet&version=4.6.0-servicing.20067.1

The question is why does this contain a newer version rather than the 4.6.1 version that was relaesed? @Anipik @ericstj System.Security.Cryptography appears newer in the shared framework vs. what was released on nuget.org (the last time the package was built). Is this the expected behavior?

Reply:
CC also @safern @joperezr in case this looks familiar. Nothing has changed in that project recently, but maybe related to https://github.com/dotnet/corefx/pull/41795 or another similar change?

Reply:
> Different versions of System.Security.Cryptography.Cng.dll are landing in both Microsoft.NetCore.App (product v3.0.3.0, file v4.700.20.6701) and Microsoft.AspNetCore.App (product v3.0.1.0, file v4.700.19.51609)

Those AssemblyVersions are surprising - corefx claims it should be `4.3.2.1`: https://github.com/dotnet/corefx/blob/a4238c9819075be9bff74e009da802ce47233f38/src/System.Security.Cryptography.Cng/Directory.Build.props#L4

Reply:
> corefx claims it should be 4.3.2.1

Yep. I see 4.3.2.1 in all of the netcore app and oob packages I can find.

Reply:
Yes the assembly versions are both 4.3.2.1, I didn't include that version because it's the same for the two assemblies. Note the words "product" and "file" before "v" for version in my [earlier comment](https://github.com/dotnet/aspnetcore/pull/18364#discussion_r368390726).

Reply:
Yeah @dougbu Is right. The question is why there is a version of cng showing up in the private netcore app package even though cng is not versioning for 3.0.3.

Reply:
It seems like that's the correct version and what we're building out of 3.0.1:

https://github.com/dotnet/corefx/blob/release/3.0/src/System.Security.Cryptography.Cng/Directory.Build.props#L4

The reason because of the inconsistency is because I believe the private package that was used for the 3.0.1 build didn't contain @wtgodbe changes, maybe? Since this project is marked as `<IsNetCoreApp>true</IsNetCoreApp>` will be included as part of the private corefx package, which will contain the `4.3.2.1` version in all builds after @wtgodbe change.

Reply:
Does that imply that aspnet should not longer have a dependency on the OOB package?

Reply:
Well, they need the `OOB` package in order to get the reference assembly and being able to use APIs from it because we don't ship the reference assembly as part of the private package:

https://github.com/dotnet/corefx/blob/release/3.0/src/System.Security.Cryptography.Cng/Directory.Build.props#L8

Because we set `<IsNetCoreAppRef>false</IsNetCoreAppRef>`. This is most likely because another library in corefx depends from it at runtime, but we don't want to make this fully inbox.

Reply:
@safern what do you suggest?

Reply:
/cc @ericstj

Reply:
The reason we didn't hit this in 3.1 may be that 3.0.1 (when we rev'd `cng`) lined up with 3.1.0, when corefx built all of their OOB packages. So there is no `3.1.1` version of `cng`.

Reply:
Lower version in package is expected for nothing that overlaps with package and shared framework. Binaries only the shared framework will always rebuild so they will have updated versions. We only ship packages when we have relevant fixes. As a result I’d expect the AspNet to be in a state where version from packages is equal or less than the shared framework.  In this case AspNet should never be including the file. It sounded to me like @dougbu was indicating the check was `=` vs `!=` instead of `<=` vs `>` as it should be.

Reply:
I believe our check is done using assembly resolution but I'm not positive. @JunTaoLuo do you remember the details?

Reply:
I guess if you're relying on conflict resolution + platform manfiest to make this decision its technically correct for ASP.NET to carry this if it's shared framework is meant to run on the lowest shared framework version of NETCore.App.  Imagine 3 versions 1 (rtm), 2 (serviced and shipped in package), 3 (serviced but no package shipped).  The platform manifest for NETCore.App will only have version 1, since apps targeting the shared framework need to be able to run on the lowest version.  If an app (or a higher-level shared framework in this case) reference the package version 2, they'll get it app-local, since it is newer than the first shipped version.  Even if we ship version 3, its not OK for an app to omit the local copy, since it doesn't target version 3, it only targets RTM.  

The difference here, is that ASP.NET is not an App, it's a shared framework.  Unlike an app which targets a TFM and a min version of a shared framework X.Y.**0**, our shared frameworks ship together and target specific versions X.Y.**Z**.  As such they shouldn't use the X.Y.0 platform manifest for conflict resolution, they should use the content of X.Y.Z.  /CC @Pilchie @nguerrera

Reply:
AFAIK, we don't have a way to get the manifest for a patch today. cc @dsplaisted.

Reply:
I haven't caught up on everything here, but how is it that this wasn't an issue in earlier 3.0.x patches?

Reply:
I think it's because CoreFx built every package every time in the previews, and in 3.1.0/3.1.1. This is the first build that comes after a build where corefx patched a package, then didn't build it in the subsequent build

Reply:
Okay so what are the next steps here then? It sounds like this is a bit of a major shift?

Reply:
I think there's an item group where ASP.NET could feed in a newer manifest. Looking.

Reply:
There's [PackageConflictPlatformManifests](https://github.com/dotnet/sdk/search?q=PackageConflictPlatformManifests&unscoped_q=PackageConflictPlatformManifests)

I think if the ASP.NET build can acquire a manifest representative of 3.0.1 and put it in there, it would work if I've understood Eric correctly.

Reply:
And where would we acquire a manifest for 3.0.1?

Reply:
Why would 3.0.1 be the right one to compare against here?

Reply:
Sorry, not 3.0.1, 3.0.(trying_to_build)

Reply:
I don't know where you would acquire it, at one point there was talk of putting it in the runtime pack, I don't know if it's in there. If not, you'd have to work out with corefx how to acquire it.

Maybe we could do something more tactical here to exclude the specific file that is causing a problem, though?

Reply:
Ah so we would need 3.0.3?

Reply:
Yeah, just the latest version that is being pulled in.

Reply:
How about core-setup produces a new non-Shipping System.System.Security.Cryptography.Cng package, perhaps w/ a new name? AspNetCore could depend on that and get the 3.0.3 assembly from it.

Reply:
@dougbu For a tactical fix, you don't even need to do that...you can just pull it from the shared framework, right?

Reply:
Not exactly, from what I can tell, it's being pulled in transitively. So we'd probably have to hack something to filter it out.

Reply:
You do pull in the rid-specific implementation packages versions,
https://github.com/dotnet/aspnetcore/blob/release/3.0/eng/Version.Details.xml#L388

Reply:
Hmm, that may be what we're doing already. @JunTaoLuo could you confirm at least a few of the assemblies we build use the 3.0.3 version of CNG in their references? If that's the case already, just need to add the new assembly to the test i.e. start _expecting_ it to show up in Microsoft.AspNetCore.App.

Reply:
I can confirm that Kestrel.Core.dll references 4.3.2.1 of Cryptography.Cng.dll. But I don't know how to tell if it's the 3.0.3 version of CNG or 3.0.1 version of it.
![image](https://user-images.githubusercontent.com/2030323/72855360-12564680-3c6c-11ea-8dbf-1d0d2187453e.png)

Just so that we're on the same page. In this case, we want to include Cryptography.Cng in our shared framework in addition to it being in the NETCore.App shared framework? I believe that in this case, we'd have two  assemblies with the same AssemblyVersion but different AssemblyFileVersion in the two shared frameworks?

Reply:
Or, am I misunderstanding the tactical fix?

Reply:
Is the tactical fix supposed to exclude Cryptography.Cng from AspNetCore.App instead?

Reply:
Can we meet tomorrow and resolve this? @dougbu @nguerrera @mmitche

Reply:
Sent a mtg request&hellip;

Reply:
Sounds good, thanks

Reply:
Tactical fix is to exclude Cng (or just keep it at the old version so long as it doesn't show up as a downgrade).

Reply:
Why don't we have the same issue with `System.Drawing.Common`? That package was patched in 3.0.1 & 3.0.2, but not in 3.0.3

Reply:
Resolution: Pin this package at 4.6.0 here and in dotnet/extensions. May need to consider source build thingies in the future ☹️

For immediate future, we'll stick w/ the test telling us about new oddities and pin as needed (when we intend to ship the implementation). That is, do not check in a rule that keeps everything we don't expect out of our slice of the shared framework -- especially in 'master' branch.

Side note: We need this package dependency to get the ref/ assembly. Pinning is fine because it's already in the .NET Core shared framework. May want to do the same for other shared framework packages that are
1. transitively referenced from ASP.NET Core assemblies in our shared framework
2. **not** in the .NET Core targeting pack
3. but **are** in the .NET Core shared framework.

See https://github.com/dotnet/runtime/search?q=isnetcoreappref&unscoped_q=isnetcoreappref for list of packages meeting (2) and (3). Still need to check (1)&hellip;

Reply:
FYI, we [list](https://github.com/dotnet/aspnetcore/blob/bd11e9503c35094d56814f6399c58c77f4220f71/eng/SharedFramework.External.props#L106-L114) `@(_CompilationOnlyReference)` items for only a small subset of the packages meeting (2) and (3) above. That _should_ mean we only need to pin the following before our next patch release:
- Microsoft.Win32.Registry
- System.Security.Cryptography.Cng
- System.Security.Principal.Windows

[List](https://github.com/dotnet/aspnetcore/blob/c3acdcac86dad91c3d3fbc3b93ecc6b7ba494bdc/eng/SharedFramework.External.props#L106-L114) looks the same in 'release/3.1' branch, implying pins should be similar though minor versions may be different (patch version should be `.0`).